src/V3RandomizeMethod.cpp

```diff
@@ -0,0 +1,171 @@
+            // If member is rand and of class type, mark its class
+            if (VN_IS(memberp, Var) && VN_CAST(memberp, Var)->isRand()) {
+                if (auto* classRefp = VN_CAST(memberp->dtypep(), ClassRefDType)) {
+                    auto* classp = classRefp->classp();
+                    markMembers(classp);
+                    markDerived(classp);
+                    classRefp->classp()->user1(true);
+                }
+            }
+        }
+    }
+
+    void markDerived(AstClass* nodep) {
+        if (m_baseToDerivedMap.find(nodep) != m_baseToDerivedMap.end()) {
+            for (auto* classp : m_baseToDerivedMap.at(nodep)) {
+                classp->user1(true);
+                markMembers(classp);
+                markDerived(classp);
+            }
+        }
+    }
+
+public:
+    void markAllDerived() {
+        for (auto p : m_baseToDerivedMap) {
+            if (p.first->user1()) { markDerived(p.first); }
+        }
+    }
+};
+
+//######################################################################
+// Visitor that defines a randomize method where needed
+
+class RandomizeMethodVisitor : public AstNVisitor {
+    virtual void visit(AstNode* nodep) override { iterateChildren(nodep); }
+
+    virtual void visit(AstClass* nodep) override {
+        iterateChildren(nodep);
+        if (nodep->user1()) {
+            auto* funcp = V3RandomizeMethod::declareIn(nodep);
+            auto* fvarp = VN_CAST(funcp->fvarp(), Var);
+            funcp->addStmtsp(new AstAssign(
+                nodep->fileline(), new AstVarRef(nodep->fileline(), fvarp, VAccess::WRITE),
+                new AstConst(nodep->fileline(), AstConst::WidthedValue(), 32, 1)));
+            auto* classp = nodep;
+            for (auto* classp = nodep; classp;
+                 classp = classp->extendsp() ? classp->extendsp()->classp() : nullptr) {
+                for (auto* memberp = classp->stmtsp(); memberp; memberp = memberp->nextp()) {
+                    auto* memberVarp = VN_CAST(memberp, Var);
+                    if (!memberVarp || !memberVarp->isRand()) continue;
+                    AstNode* stmtp = nullptr;
+                    auto* refp = new AstVarRef(nodep->fileline(), memberVarp, VAccess::WRITE);
+                    if (VN_IS(memberp->dtypep()->skipRefp(), BasicDType)) {
+                        stmtp = new AstStdRandomize(nodep->fileline(), refp);
+                    } else if (auto* classRefp = VN_CAST(memberp->dtypep(), ClassRefDType)) {
+                        auto* funcp = V3RandomizeMethod::declareIn(classRefp->classp());
+                        auto* callp
+                            = new AstMethodCall(nodep->fileline(), refp, "randomize", nullptr);
+                        callp->taskp(funcp);
+                        callp->dtypeFrom(funcp);
+                        stmtp = callp;
+                    } else {
+                        delete refp;
```

Comment:
This would need to be a VL_DO_DANGLING(refp->deleteTree(), refp)
However doing new's you just delete isn't nice, instead please duplicate the new AstVarRef (or move to a common function) you call under both if branches.